tools/autograd/templates/VariableType.cpp

```diff
@@ -29,8 +29,14 @@
 using namespace at;
 using namespace torch::autograd::generated;
 
+${manual_functions_h}
+
 namespace torch { namespace autograd {
 
+namespace {
+${manual_functions_cpp}
```

Comment:
If I'm understanding this correctly, manual_functions.cpp will be compiled into VariableType.cpp AND Functions.cpp.

This would increase the binary size, but probably not by a very significant amount (the compiler hopefully knows how to throw away the parts that don't get used).

What are the functions that you wanted to share between the forward implementations and Functions.cpp? Is it only a few or is it a lot of things? If we were writing straight-up C++ we would define things to be shared in a header and include the header where necessary, but the codegen does give us a lot more freedom

Reply:
It turns out that most of the forward formulas actually need the backward definition as well.
I went with this version to:
- Keep the anonymous namespace to make sure we don't expose these in other places by mistake
- Assumed that the compiler will just remove everything that's not needed indeed
- Keep the manual Functions definition close to where they are right now so that people don't get lost looking for them.

Reply:
Well, this also increases compilation time because the code now gets compiled as a part of N compilation units, where N is the number of VariableType shards we do (5?). Furthermore, modifying something in ManualFunctions.cpp will trigger a recompile of all of the VariableType shards.

I'm still wary of adding new things to the codegen. However, the codegen changes are pretty minimal, so we can make the case that this isn't too crazy of a change.

As an alternative, would something like the following work?
- Add ManualFunctions.h and ManualFunctions.cpp.
- ManualFunctions.h contains declarations for all of the functions in Functions.cpp in a new torch::autograd::generated::details namespace
- ManualFunctions.cpp contains definitions for all of the functions in Functions.cpp in torch::autograd::generated::details namespace
- `#include ManualFunctions.h` into Functions.cpp and VariableType.cpp (and just use `using torch::autograd::generated::details` inside those .cpp files).

This way, we don't expose any symbols in other places, keep the manual functions definition close to where they are now, don't need to modify the codegen, and keep a nice header/body split. The only thing that is tedious is listing declarations in ManualFunction.h (but one can probably write a script to toss out the names of all functions in Functions.cpp)

Reply:
I agree with @zou3519. Even if you don't want to put things in a cpp file and duplicate them, you should simply mark all the functions as `inline` and put them in a header (and don't transclude them manually via codegen). But it is much better to put them in a cpp file so there is only one copy floating around.

Reply:
Yes after a quick chat with Richard, I will move these out of the codegen code and just include them once.